lib/routes/recovery-codes.js

```diff
@@ -0,0 +1,177 @@
+            .then((result) => {
+              codes = result
+            })
+        }
+
+        function sendEmailNotification() {
+          return db.account(sessionToken.uid)
+            .then((account) => {
+              mailer.sendPostNewRecoveryCodesNotification(account.emails, account, {
+                acceptLanguage: request.app.acceptLanguage,
+                ip: ip,
+                location: geoData.location,
+                timeZone: geoData.timeZone,
+                uaBrowser: request.app.ua.browser,
+                uaBrowserVersion: request.app.ua.browserVersion,
+                uaOS: request.app.ua.os,
+                uaOSVersion: request.app.ua.osVersion,
+                uaDeviceType: request.app.ua.deviceType,
+                uid: sessionToken.uid
+              })
+            })
+        }
+
+        function emitMetrics() {
+          log.info({
+            op: 'account.recoveryCode.replaced',
+            uid: uid
+          })
+
+          return request.emitMetricsEvent('recoveryCode.replaced', {uid: uid})
+            .then(() => ({}))
+        }
+      }
+    },
+    {
+      method: 'POST',
+      path: '/session/verify/recoveryCode',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: {
+            code: isA.string().min(RECOVERY_CODE_LENGTH).max(RECOVERY_CODE_LENGTH).regex(HEX_STRING).required()
+          }
+        },
+        response: {
+          schema: {
+            remaining: isA.number()
+          }
+        }
+      },
+      handler(request, reply) {
+        log.begin('session.verify.recoveryCode', request)
+
+        const code = request.payload.code
+        const uid = request.auth.credentials.uid
+        const sessionToken = request.auth.credentials
+        const geoData = request.app.geo
+        const ip = request.app.clientAddress
+        let remaining, isValidCode = false
+
+        customs.check(request, 'verifyRecoveryCode')
+          .then(consumeRecoveryCode)
+          .then(verifySession)
+          .then(sendEmailNotification)
+          .then(emitMetrics)
+          .then(() => {
+            return reply({remaining: remaining})
+          }, reply)
+
+        function consumeRecoveryCode() {
+          return db.consumeRecoveryCode(uid, code)
+            .then((result) => {
+              isValidCode = true
+              remaining = result.remaining
+              if (remaining === 0) {
+                log.info({
+                  op: 'account.recoveryCode.consumedAllCodes',
+                  uid: uid
+                })
+              }
+            })
+        }
+
+        function verifySession() {
+          if (sessionToken.tokenVerificationId) {
+            return db.verifyTokensWithMethod(sessionToken.id, 'recovery-code')
+          }
+        }
+
+        function sendEmailNotification() {
+          if (! isValidCode) {
+            return
```

Comment:
Is this unreachable code?

It looks like `isValidCode` will only be `false` if `consumeRecoveryCode` rejected. But if `consumeRecoveryCode` rejected, we'd have fallen straight through to the final `reply` handler. So it seems like `isValidCode` doesn't really serve any purpose in this endpoint, unless I'm reading the code wrong?